File: src/main/java/legacy/kms/chapter02/discount/DefaultDiscountPolicy.java
package legacy.kms.chapter02.discount;

import legacy.kms.chapter02.Money;
import legacy.kms.chapter02.Screening;

import java.util.Arrays;
import java.util.List;

public abstract class DefaultDiscountPolicy implements DiscountPolicy {

    private List<DiscountCondition> conditions;

    public DefaultDiscountPolicy(DiscountCondition ... conditions) {
        this.conditions = Arrays.asList(conditions);
    }

    @Override
    public Money calculateDiscountAmount(Screening screening) {
        for(DiscountCondition each : conditions){
            if(each.isSatisfiedBy(screening)) {
                return getDiscountAmount(screening);
            }
        }
        return Money.ZERO;
    }

    abstract protected Money getDiscountAmount(Screening screening);
}

File: src/main/kotlin/kms/chapter02/discount/DefaultDiscountPolicy.kt
package kms.chapter02.discount

import kms.chapter02.Money
import kms.chapter02.Screening

internal abstract class DefaultDiscountPolicy(
    private val conditions: List<DiscountCondition>,
) : DiscountPolicy {

    constructor(vararg conditions: DiscountCondition) : this(conditions.toList())

    override fun calculateDiscountAmount(screening: Screening): Money {
        for (condition in conditions) {
            if (condition.isSatisfiedBy(screening)) {
                return getDiscountAmount(screening)
            }
        }
        return Money.ZERO
    }

    protected abstract fun getDiscountAmount(screening: Screening): Money
}
